src/core/include/lattice/hal/default/dcrtpoly-impl.h

```diff
@@ -1926,128 +1926,105 @@ DCRTPolyImpl<VecType> DCRTPolyImpl<VecType>::ScaleAndRound(
 
-                ans.m_vectors[j][ri] = curNativeValue.ModAdd(alpha, oj, mu[j]);
+                auto curAlpha{alpha};
+                if (alpha >= oj)
+                    curAlpha = alpha.Mod(oj, mu[j]);
+                ans.m_vectors[j][ri] = NativeInteger(curNativeValue).ModAddFast(curAlpha, oj);
             }
         }
         else {
-            int exp;
-            double mant           = std::frexp(nu, &exp);
-            uint64_t mantissa     = static_cast<uint64_t>(mant * (1ULL << 53));
-            uint64_t exponent     = static_cast<uint64_t>(1ULL << (exp - 53));
-            DoubleNativeInt alpha = Mul128(mantissa, exponent);
-
-            for (usint j = 0; j < sizeO; j++) {
-                DoubleNativeInt curValue = 0;
-
-                const NativeInteger& oj                                  = paramsOutput->GetParams()[j]->GetModulus();
-                const std::vector<NativeInteger>& tOSHatInvModsDivsModoj = tOSHatInvModsDivsModo[j];
-
-                for (usint i = 0; i < sizeI; i++) {
+            auto alpha = static_cast<DoubleNativeInt>(nu);
+            for (size_t j = 0; j < sizeO; ++j) {
+                const auto& tOSHatInvModsDivsModoj = tOSHatInvModsDivsModo[j];
+                DoubleNativeInt curValue{0};
+                for (size_t i = 0; i < sizeI; ++i) {
                     const NativeInteger& xi = m_vectors[i + inputIndex][ri];
                     curValue += Mul128(xi.ConvertToInt(), tOSHatInvModsDivsModoj[i].ConvertToInt());
                 }
-
                 const NativeInteger& xi = m_vectors[outputIndex + j][ri];
                 curValue += Mul128(xi.ConvertToInt(), tOSHatInvModsDivsModoj[sizeI].ConvertToInt());
 
-                const NativeInteger& curNativeValue =
-                    NativeInteger(BarrettUint128ModUint64(curValue, oj.ConvertToInt(), modoBarretMu[j]));
+                const NativeInteger& oj = paramsOutput->GetParams()[j]->GetModulus();
+                auto&& curNativeValue   = BarrettUint128ModUint64(curValue, oj.ConvertToInt(), modoBarretMu[j]);
 
                 ans.m_vectors[j][ri] =
-                    curNativeValue.ModAddFast(BarrettUint128ModUint64(alpha, oj.ConvertToInt(), modoBarretMu[j]), oj);
+                    NativeInteger(curNativeValue).ModAddFast(BarrettUint128ModUint64(alpha, oj.ConvertToInt(), modoBarretMu[j]), oj);
             }
         }
     }
-
     return ans;
 }
 
 #else
     #pragma omp parallel for
-    for (usint ri = 0; ri < ringDim; ri++) {
+    for (usint ri = 0; ri < ringDim; ++ri) {
         double nu = 0.5;
-
-        for (usint i = 0; i < sizeI; i++) {
+        for (size_t i = 0; i < sizeI; ++i) {
+            // possible loss of precision if modulus greater than 2^53 + 1
             const NativeInteger& xi = m_vectors[i + inputIndex][ri];
-            nu += tOSHatInvModsDivsFrac[i] * xi.ConvertToInt();
+            nu += tOSHatInvModsDivsFrac[i] * xi.ConvertToDouble();
         }
-
         if (!is64BitOverflow(nu)) {
```

Comment:
This is encountered when NATIVEINT == 32, NATIVEINT == 128, or !defined(HAVE_INT128) && NATIVEINT == 64.

There are various subtle issues with current implementation if NATIVEINT == 32 or NATIVEINT == 128. Should we create templated function that performs correct overflow check dependent upon BasicIntegr type?

Reply:
The current assumption is that that BFV should only be used for NATIVE_SIZE=64 (both when 128-bit int is available and not)